### ISSUE TITLE
Possibility to change exception being used

### ISSUE THREAD

**Issue by andrzejkupczyk** [thunderer/Platenum#6]
Since the Platenum could be used in a business-model it would be great if it was possible to change the exception being used on a package based level. Platenum's exceptions may not reflect 100% given domain and cause leaking implementation details of a generic subdomain. What do you think, @thunderer?

Alternatively, maybe it would be better to throw "raw" SPL exceptions instead?

**Comment by thunderer** [comment]
Hi @andrzejkupczyk, thanks for reporting the issue! Can you provide examples of what changes would you like to see in the library, especially from the end-user perspective? I get the "leaking implementation details" issue, but I'm not sure what do you mean by "change the exception on package level". Do you want to abstract the way the exceptions are thrown? If so, how would you expect this to be implemented here?

Platenum throws instances of one exception, which represents multiple "sanity checks" about the way enumerations should be used - that's why I didn't split it into separate types. This exception explicitly says the code needs to be fixed and there is no recovery, like empty members list or missing `resolve()` method. Only `::fromIllegalValue()` looks like something that would benefit from separate exception type to be caught and handled, because invalid member value is one of the recoverable errors when for example validating user input.

Let me know what do you think! :)

**Comment by andrzejkupczyk** [comment]
Thank you for your prompt reply :1st_place_medal:

Throwing instances of one exception, as you have rightly pointed out, may be fine. What I meant is a possibility to indicate an exception class that Platenum throws instances of. From the end-user point of view I would like to see something like this:

```php
<?php

namespace Model\User;

final class AccountStatus extends AbstractConstantsEnum
{
    private const ACTIVE = 1;
    private const INACTIVE = 2;

    protected static $exceptionClass = InvalidAccountStatus::class;
}

AccountStatusEnum::fromValue(3); // throws InvalidAccountStatus exception
```

The problem with this approach however is that currently thrown exceptions are mixed up in some way. Part of them comes from helper methods that handles building the enum failure exceptions (runtime exceptions?, e.g. "fromMissingDocblock"). Other relates directly to domain rules (logic exceptions?).

I don't know if you agree but Platenum always throws `PlatenumException` (`LogicException`) with messages that may be unclear in a given domain / context.

I could probably create a decorator that hides original enum's interface and handles those exceptions but maybe there are enhancements that could be done within Platenum itself? Or maybe all of this is not a real problem? :wink:

**Comment by thunderer** [comment]
Hi @andrzejkupczyk, I had drafts of a few variants of the implementation shortly after your last comment, but finally got a spare moment to refine it and create initial version of PR #8. The idea is that custom exception classes will be configurable through dedicated properties on the enumeration class:

```php
final class YourEnum
{
    use EnumTrait;

    private static $invalidMemberExceptionClass = YourMemberException::class;
    private static $invalidValueExceptionClass = YourValueException::class;

    // members, etc.
}
```

These properties are then remapped to an internal `$exceptions` mapping in `EnumTrait` and checked right before throwing the exception - if there is no override then regular `PlatenumException` is thrown. The passed parameters are the same, the custom exception classes receive them in default constructor instead of named ones. Right now I'm just wondering which cases should be overridable and exposed through that mechanism.

I also thought about introducing `PlatenumExceptionInterface` to mimic the way `PlatenumException` is implemented and allow a single class to replace all cases, but IMHO this would constrain the overall design without any direct benefit. In your example, you have a single `$exceptionClass` property - did you have something like that in mind? Let's continue the discussion. :)

**Comment by gjm** [comment]
@thunderer @andrzejkupczyk If I may step in here, I would suggest a different approach: having a set of methods in `EnumTrait` (or maybe event extract them to an `EnumExceptions` trait) that throw each of the supported exceptions:
```php
protected static function illegalValueException(string $fqcn, $value)
{
    throw PlatenumException::fromIllegalValue($fqcn, $value);
}

protected static function mismatchedClassException(string $fqcn, string $other)
{
    throw static PlatenumException::fromMismatchedClass($fqcn, $other);
}
...
```

In the code, replace references of `throw PlatenumException::fromIllegalValue(...)` with `static::illegalValueException(...)`.

This way, in each subclass one could override the relevant methods and could build and throw the exceptions for each case. That way we would have total freedom when throwing the exceptions. Maybe we want to keep the exception chain... maybe not... maybe throw a standard SPL exception... maybe a domain specific one... whatever...

Another alternative is just returning the exception instance in each method:
```php
protected static function illegalValueException(string $fqcn, $value): \Exception
{
    return PlatenumException::fromIllegalValue($fqcn, $value);
}
```

and, in code: `throw static::illegalValueException($fcqn, $value)`

**Comment by thunderer** [comment]
@gjm Thanks again for your contributions, I reimplemented #8 as per your comment. Please review the PR and let me know what you think. This is an excellent idea that not only simplifies the internal design but also gives much more flexibility for the end-users.

Could you please also look at #9 and #12? It would be good to have more than one pair of eyes on them. :)

**Comment by gjm** [comment]
@thunderer #8 seems OK to me. Wouldn't see any harm in extending the same behavior for the `valueToMember` and `memberToValue` methods but the way it is now seems to serve the purpose.

As for #9 and #12 I don't use Psalm and I'm not really familiar with Doctrine either but, just looking at them, both seem fine to me too.

**Comment by andrzejkupczyk** [comment]
I like it, thank you both. I'm really looking forward to the next release :clap:

**Comment by thunderer** [comment]
@gjm @andrzejkupczyk good news, I just merged #8 - in addition to the last PR state I added a safeguard that still throws a default exception when none is thrown in the original method.

Thanks for all your help and using Platenum! I'm closing this issue, but let me know if there is anything else I can help you with.

**Comment by thunderer** [comment]
@gjm @andrzejkupczyk I had a pretty active Sunday and just released `v0.3` - I hope [the changelog will make you happy](https://github.com/thunderer/Platenum/releases/tag/v0.3). :) Custom exceptions, callback source, Doctrine type, and many more... let me know what do you think.
